File: bench/callbacks/callbacks.go
// Functions as values: closures, composition, and list helpers that call back
// into a function for every element.
package main

import "fmt"

func compose(f, g func(int) int) func(int) int {
	return func(x int) int { return f(g(x)) }
}

func mapSlice(fn func(int) int, values []int) []int {
	out := []int{}
	k := 0
	for k < len(values) {
		out = append(out, fn(values[k]))
		k = k + 1
	}
	return out
}

func filterSlice(fn func(int) bool, values []int) []int {
	out := []int{}
	k := 0
	for k < len(values) {
		if fn(values[k]) {
			out = append(out, values[k])
		}
		k = k + 1
	}
	return out
}

func reduceSlice(fn func(int, int) int, values []int, initial int) int {
	acc := initial
	k := 0
	for k < len(values) {
		acc = fn(acc, values[k])
		k = k + 1
	}
	return acc
}

func main() {
	double := func(x int) int { return x * 2 }
	increment := func(x int) int { return x + 1 }
	step := compose(increment, double)

	xs := []int{}
	i := 0
	for i < 2000 {
		xs = append(xs, i)
		i = i + 1
	}

	total := 0
	round := 0
	for round < 1600 {
		mapped := mapSlice(step, xs)
		kept := filterSlice(func(x int) bool { return x%3 == 0 }, mapped)
		total = total + reduceSlice(func(acc, x int) int { return acc + x }, kept, 0)
		round = round + 1
	}
	fmt.Println(total)
}

File: bench/callbacks/callbacks.js
// Functions as values: closures, composition, and list methods that call back
// into a function for every element.
function compose(f, g) {
    return (x) => f(g(x));
}

const double = (x) => x * 2;
const increment = (x) => x + 1;
const step = compose(increment, double);

const xs = [];
let i = 0;
while (i < 2000) {
    xs.push(i);
    i = i + 1;
}

let total = 0;
let round = 0;
while (round < 1600) {
    const mapped = xs.map(step);
    const kept = mapped.filter((x) => x % 3 === 0);
    total = total + kept.reduce((acc, x) => acc + x, 0);
    round = round + 1;
}
console.log(total);
